lib/_http_client.js

```diff
@@ -199,7 +199,13 @@ function ClientRequest(input, options, cb) {
   const host = optsWithoutSignal.host = validateHost(options.hostname, 'hostname') ||
                                         validateHost(options.host, 'host') || 'localhost';
 
-  const setHost = (options.setHost === undefined || Boolean(options.setHost));
+  const setHost = options.setHost !== undefined ?
+    Boolean(options.setHost) :
+    options.setDefaultHeaders !== false;
+
+  this._removedConnection = options.setDefaultHeaders === false;
+  this._removedContLen = options.setDefaultHeaders === false;
+  this._removedTE = options.setDefaultHeaders === false;
```

Comment:
It's not clear what these are for or why they are needed.

```suggestion
  this._removedConnection = options.setDefaultHeaders === false;
  this._removedContLen = options.setDefaultHeaders === false;
  this._removedTE = options.setDefaultHeaders === false;
```

Reply:
These lines are the key lines of this change - it doesn't work otherwise.

They change the internal 'should we automatically add these headers' state for Connection/Content-Length/Transfer-Encoding - if `false`, the headers are added automatically. By default they're set to `false` [here](https://github.com/nodejs/node/blob/fd9a79df7dcf63990ff15c99433d6d86c95b794e/lib/_http_outgoing.js#L128-L130) in OutgoingMessage, and in the current codebase they're only set to `true` [here](https://github.com/nodejs/node/blob/fd9a79df7dcf63990ff15c99433d6d86c95b794e/lib/_http_outgoing.js#L853-L866) if you call `removeHeader(...)` for the corresponding header (as noted in the PR description - it's not possible to call this method in the raw header case).

These lines mean that they're set to `true` up front (overriding the OutgoingMessage default and disabling the automatic inclusion of these headers, as if the 3 `removeHeader(...)` calls had been made) if `setDefaultHeaders: false` is passed in the options.

Does that make sense @jasnell?

Reply:
Ugh, yeah I forgot about that in `OutgoingMessage` these are existing fields. We should see about changing these at some point to internal private fields.